instance/models/instance.py

```diff
@@ -455,29 +461,53 @@ def run_playbook(self):
         """
         Run a playbook against the instance active servers
         """
-        with open_repository(self.ansible_source_repo_url,
-                             ref=self.configuration_version) as configuration_repo:
-            playbook_path = os.path.join(configuration_repo.working_dir, 'playbooks')
-            requirements_path = os.path.join(configuration_repo.working_dir, 'requirements.txt')
-
-            self.logger.info('Running playbook "%s/%s"...', playbook_path, self.ansible_playbook_filename)
-
-            log_lines = []
-            with ansible.run_playbook(
-                requirements_path,
-                self.inventory_str,
-                self.vars_str,
-                playbook_path,
-                self.ansible_playbook_filename,
-                username=settings.OPENSTACK_SANDBOX_SSH_USERNAME,
-            ) as processus:
-                for line in processus.stdout:
-                    line = line.decode('utf-8').rstrip()
-                    self.logger.info(line)
-                    log_lines.append([line.rstrip()])
-
-        self.logger.info('Playbook run completed')
-        return log_lines
+        attempt = 0
+        while attempt < self.attempts:
+            with open_repository(self.ansible_source_repo_url,
+                                 ref=self.configuration_version) as configuration_repo:
+                playbook_path = os.path.join(configuration_repo.working_dir,
+                                             'playbooks')
+                requirements_path = os.path.join(configuration_repo.working_dir,
+                                                 'requirements.txt')
+
+                log = ('Running playbook "{path}/{name}" for instance {instance}, '
+                       'attempt {attempt} of {attempts}:').format(
+                           path=playbook_path,
+                           name=self.ansible_playbook_name,
+                           instance=self,
+                           attempts=self.attempts,
+                           attempt=attempt + 1)
+                self.logger.info(log)
+
+                log_lines = []
+                with ansible.run_playbook(
+                    requirements_path,
+                    self.inventory_str,
+                    self.vars_str,
+                    playbook_path,
+                    self.ansible_playbook_filename,
+                    username=settings.OPENSTACK_SANDBOX_SSH_USERNAME,
+                ) as processus:
+                    output = itertools.chain(
+                        zip(itertools.repeat(logging.INFO), processus.stdout),
+                        zip(itertools.repeat(logging.ERROR), processus.stderr)
+                    )
```

Comment:
Iterating over both stdout and stderr at the same time is notoriously prone to deadlocks.  This code will iterate over stdout first, so if the output buffer for stderr runs full while the process is still running, this will simply block.  It's probably unlikely to actually happen in this particular case, but we should code defensively against this anyway.  One way to do so is using the `communicate()` method of the process, but this doesn't allow forwarding the output in real time.  So some asynchronous mechanism is needed, be it `select.poll()`, threads or using Python 3.4's new `asyncio` module.